include/boost/geometry/algorithms/detail/envelope/multipoint.hpp

```diff
@@ -0,0 +1,362 @@
+            }
+        }
+    }
+
+    template
+    <
+        typename Constants,
+        typename Iterator,
+        typename LatitudeLess,
+        typename CoordinateType
+    >
+    static inline void get_min_max_latitudes(Iterator const first,
+                                             Iterator const last,
+                                             LatitudeLess const& lat_less,
+                                             bool has_south_pole,
+                                             bool has_north_pole,
+                                             CoordinateType& lat_min,
+                                             CoordinateType& lat_max)
+    {
+        if (has_south_pole && has_north_pole)
+        {
+            lat_min = Constants::min_latitude();
+            lat_max = Constants::max_latitude();
+        }
+        else if (has_south_pole)
+        {
+            lat_min = Constants::min_latitude();
+            lat_max = std::max_element(first, last, lat_less)->second;
+        }
+        else if (has_north_pole)
+        {
+            lat_min = std::min_element(first, last, lat_less)->second;
+            lat_max = Constants::max_latitude();
+        }
+        else
+        {
+            std::pair<Iterator, Iterator> min_max_latitudes
+                = boost::minmax_element(first, last, lat_less);
+
+            lat_min = min_max_latitudes.first->second;
+            lat_max = min_max_latitudes.second->second;
+        }
+    }
+
+public:
+    template <typename MultiPoint, typename Box>
+    static inline void apply(MultiPoint const& multipoint, Box& mbr)
+    {
+        typedef typename coordinate_type<MultiPoint>::type coordinate_type;
+        typedef lon_lat_pair<coordinate_type> pair_type;
+
+        typedef math::detail::constants_on_spheroid
+            <
+                coordinate_type,
+                typename coordinate_system<MultiPoint>::type::units
+            > constants;
+
+
+        if (boost::empty(multipoint))
+        {
+            return;
+        }
+
+        if (boost::size(multipoint) == 1)
+        {
+            return dispatch::envelope
+                <
+                    typename boost::range_value<MultiPoint>::type
+                >::apply(range::front(multipoint), mbr);
+        }
+
+        // analyze the coordinates and put the non-pole ones in the
+        // coordinates vector
+        std::vector<pair_type> coordinates;
+        bool has_north_pole(false), has_south_pole(false);
+
+        analyze_point_coordinates<constants>(multipoint,
+                                             has_south_pole, has_north_pole,
+                                             std::back_inserter(coordinates));
+
+        coordinate_type lon_min, lat_min, lon_max, lat_max;
+        if (coordinates.size() == 1)
+        {
+            // we have one non-pole point and at least one pole point
+            lon_min = lon_max = coordinates.begin()->first;
```

Comment:
The `lon_lat_pair` is now gone. I am using points instead.
The code should be more readable now.